### ISSUE TITLE
walk.lua:295: Humanoid (Stripped Female Patient 3) without door animations trying to walk through door (from gene ral_diag to corridor)

### ISSUE THREAD

**Issue by ghost** [CorsixTH/CorsixTH#258]
The game asked me if I want to recover and dumped the below logs. After capturing the logs, I said 'yes' and it crashed. Latest version of Corsix.

```
Running: The timer handler.
A stack trace is included below, and the handler has been disconnected.
C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:295: Humanoid (Stripped
Female Patient 3) without door animations trying to walk through door (from gene
ral_diag to corridor)
stack traceback:
        [C]: in function 'error'
        C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:295: in function
 'action_walk_raw'
        C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:179: in function
 'timer_function'
        C:\Program Files\CorsixTH\Lua\entity.lua:166: in function 'tick'
        C:\Program Files\CorsixTH\Lua\world.lua:1035: in function 'onTick'
        C:\Program Files\CorsixTH\Lua\app.lua:838: in function 'dispatch'
        C:\Program Files\CorsixTH\Lua\app.lua:739: in function <C:\Program Files
\CorsixTH\Lua\app.lua:734>

Warning: No event handler for timer
```

**Comment by mugmuggy** [comment]
I've come across this one a few times in testing other things, but hadn't devoted anytime to it. Whilst brainstorming another issue I believe I stumbled on at least one root cause. Sort of related to 1233 but a little different.

I believe what occurs in this instance is the doctor/patient are finished with the general diagnosis. Doctor is leaving for another room. Patient is sent to dress as normal, patient is leaving the room, but again, a handyman is using the door, the patient then must be queuing, and loosing the is_leaving flag state.

The handyman then exits the room. And the following code dealing with patients isn't guarded for handymen

https://github.com/CorsixTH/CorsixTH/blob/627a5e251e07d6a6c4d4cce1af5df523be0a62cd/CorsixTH/Lua/room.lua#L560-L578 

The shouldHavePatientReenter checks for anything in the action queue with is_leaving = true and theoretically guard against the patient_needs_to_reenter flag being set to true. When I got this error I had just inserted some logic to record everytime, the above snippet of code was called and that flag was true for a handyman. I assume the easiest fix is to check for handyman on the outermost if. When I have some more time I will try and replicate it based on this. I had replicated it once or twice before manually (in many many attempts) but didn't understand the potential why and it still could be timing related for the patients state.

**Comment by TheCycoONE** [comment]
In the forums, there was mention of a number of places in room.lua that don't appear to appropriately filter for staff type. Unfortunately they didn't bring the list over here, and I seldom check the forums, but you seem to be confirming their suspicion.

I wonder if there shouldn't be subclasses of staff.lua for each of the types, and more checks for `class.is(humanoid, Doctor)` or `class.is(humanoid, Nurse)` instead of `class.is(humanoid, Staff)` (An awful lot of Staff.lua is specific to Doctors as it is)

**Comment by lewri** [comment]
I can't replicate this one myself... though I imagine it does still exist.
I've tried dropping a handyman in the room to leave before a patient gets dressed again -- but it didn't cause any issues.

**Comment by lewri** [comment]
I wonder if this one is somewhat related to the surgeon bug in #1933

**Comment by mugmuggy** [comment]
It isn't, this is NPC. This issue shouldn't occur any more from my understanding. This should be similar to #293 and the other reason it used to occur was during evacuations during epidemic (fix I believe in #1476).

**Comment by lewri** [comment]
@TheCycoONE are you happy this one is likely resolved by 1476?
